Authorization/JAXB-API for Jenkins #55, #57

gnl42 opened the pull request:
Added "crumbissuer" authentication. Do we still need the old authentication code?

Fixed missing JAXB-API complaint 
(https://www.vogella.com/blog/eclipse-rcp-java-11-jaxb/)

wimjongman commented on mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:
```diff
@@ -70,11 +74,58 @@ protected String baseUrl() {
+	}
+
+	@Deprecated
+	private void legacyAuthentication(IOperationMonitor monitor, UserCredentials credentials)
```
You can delete this.

wimjongman commented on mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:
```diff
@@ -109,14 +160,15 @@ protected void authenticate(IOperationMonitor monitor) throws IOException {
 	}
 
-	public HttpResponse executeAuthenticationRequest(IOperationMonitor monitor, UserCredentials credentials,
+	@Deprecated
```
Please drop.

gnl42 replied:
Alas, the Hudson/Jenkins servers used for the unit tests doesn't support the new way, so I put them back

wimjongman commented on mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:
```diff
@@ -70,11 +74,58 @@ protected String baseUrl() {
+
+	}
+
+	@Deprecated
```
You can drop it.

wimjongman commented on mylyn.builds/org.eclipse.mylyn.hudson.core/META-INF/MANIFEST.MF:
```diff
@@ -4,14 +4,21 @@ Bundle-Name: %Bundle-Name
  org.eclipse.mylyn.commons.repositories.core;bundle-version="3.26.0",
  org.eclipse.mylyn.commons.repositories.http.core;bundle-version="3.26.0",
- org.eclipse.mylyn.hudson.core;bundle-version="3.26.0"
+ org.eclipse.mylyn.hudson.core;bundle-version="3.26.0",
```
Can't you use import package instead of require bundle? Also, are all of these required?

wimjongman commented on mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:
```diff
@@ -254,7 +327,7 @@ private boolean hasValidatAuthenticationState() {
 		if (cookies != null) {
 			for (Cookie cookie : cookies) {
-				if ("JSESSIONID".equals(cookie.getName())) { //$NON-NLS-1$
+				if ("JSESSIONID".equals(cookie.getName()) || cookie.getName().startsWith("JSESSIONID")) { //$NON-NLS-2$
```
Put in JSESSIONID in a constant?

wimjongman reviewed (changes requested):
Thanks, George! 

This looks good. I found a couple of nits.

ruspl-afed commented on mylyn.builds/org.eclipse.mylyn.hudson.core/META-INF/MANIFEST.MF:
```diff
@@ -4,14 +4,18 @@ Bundle-Name: %Bundle-Name
 Bundle-Vendor: %Bundle-Vendor
-Require-Bundle: jakarta.xml.bind;bundle-version="2.3.3",
- org.eclipse.core.runtime;bundle-version="0.0.0",
+Require-Bundle: org.eclipse.core.runtime;bundle-version="0.0.0",
```
Please keep alphabetic order for dependencies

ruspl-afed commented on org.eclipse.mylyn-target/mylyn-e4.26.target:
```diff
@@ -71,6 +71,22 @@
+				    <version>2.3.3</version>
+				</dependency>
+				<dependency>
+				    <groupId>org.glassfish.hk2</groupId>
```
Why do we need this one?

gnl42 replied:
I have no idea how it's all wired together. Maybe https://www.vogella.com/blog/eclipse-rcp-java-11-jaxb/ will make more sense to you?

ruspl-afed replied:
Yes, true, seems to be required to activate plain java services expected by JAXB impl with OSGi. We can get this bundle from https://download.eclipse.org/tools/orbit/downloads/drops/R20221123021534/ as well.

gnl42 replied:
Hmm, when I moved it to orbit, the build breaks with:
Error:  Cannot resolve project dependencies: Error:    Software being installed: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier Error:    Missing requirement: org.eclipse.mylyn.commons.ui 3.26.0.qualifier requires 'osgi.bundle; org.apache.commons.lang [2.6.0,3.0.0)' but it could not be found Error:    Cannot satisfy dependency: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier depends on: org.eclipse.equinox.p2.iu; org.eclipse.mylyn.commons.ui 0.0.0

I also tried the 2.3.0 version, but it has similar, but different, problems

gnl42 replied:
Tried to change it to commons.lang3 but it's not available from orbit(?)

ruspl-afed replied:
OMG. Let's postpone this improvement, I'd rather have a closer look on our .target later.

ruspl-afed commented on org.eclipse.mylyn-target/mylyn-e4.26.target:
```diff
@@ -71,6 +71,22 @@
+				    <version>2.3.3</version>
+				</dependency>
+				<dependency>
+				    <groupId>jakarta.xml.bind</groupId>
```
we have this in the target already from `https://download.eclipse.org/tools/orbit/downloads/drops/R20221123021534`

ruspl-afed commented on org.eclipse.mylyn-target/mylyn-e4.26.target:
```diff
@@ -71,6 +71,22 @@
 		<location includeDependencyDepth="infinite" includeDependencyScopes="compile,provided,runtime" includeSource="true" missingManifest="generate" type="Maven">
 			<dependencies>
+				<dependency>
+				    <groupId>com.sun.xml.bind</groupId>
```
Would be better to use one from `https://download.eclipse.org/tools/orbit/downloads/drops/R20221123021534`

gnl42 replied:
Hmm, when I moved it to orbit, the build breaks with:
`Error:  Cannot resolve project dependencies:
Error:    Software being installed: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier
Error:    Missing requirement: org.eclipse.mylyn.commons.ui 3.26.0.qualifier requires 'osgi.bundle; org.apache.commons.lang [2.6.0,3.0.0)' but it could not be found
Error:    Cannot satisfy dependency: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier depends on: org.eclipse.equinox.p2.iu; org.eclipse.mylyn.commons.ui 0.0.0`

gnl42 replied:
Never mind. It's only the source that's available from orbit. Sticking to the maven version for now

gnl42 replied:
Orbit searches are weird. I did find the right one, but it has the same problem.
I also tried the 2.3.0 version, but it has similar, but different, problems

gnl42 commented:
Hi Wim, 
  I made the changes and had to rework the code a bit to fix some issues I came across after some more testing.

The build framework is weird, I don't understand what the intent was in its design.
CommonHttpClient has an "authenticated" flag that is good for a session, yet has a ThreadLocal variable "context" that is used to store the token. Side effect of this is that each thread had to authenticate rather than just doing it once for the session.

One issue I'm unable to solve is how use the JSESSION id to authenticate when running a build using ones password.
It works fine when using a API token, but according to all the documentation I've come across it should be doable in both cases???

I've been testing it against my local Jenkins instance, and against https://ci.eclipse.org/mylyn/job/github-mylyn/.
If someone who is able to login to ci.eclipse.org could test the code I would feel more comfortable.

Also, I haven't seen any sign that the crumb is needed. Could be my jenkins hasn't been set up fully.

BeckerFrank commented:
> Also, I haven't seen any sign that the crumb is needed. Could be my jenkins hasn't been set up fully.

I have finally remember the password so that I could reactivate the Hudson and Jenkins test instances of the http://mylyn.org domain. Maybe that helps.

gnl42 commented:
> > Also, I haven't seen any sign that the crumb is needed. Could be my jenkins hasn't been set up fully.
> 
> I have finally remember the password so that I could reactivate the Hudson and Jenkins test instances of the http://mylyn.org domain. Maybe that helps.

If that is the same instances that the unit tests are using (e.g.: https://mylyn.org/hudson-3.3.3, http://mylyn.org/jenkins-2.32.3/) it doesn't help.

Neither of them accept the ''crumbIssuer/api/json" request

BeckerFrank commented:
Maybe the versions are to old to support that or we have never configure this.

gnl42 commented:
@BeckerFrank 
> Maybe the versions are to old to support that or we have never configure this.

Looks like 2.32.3 was released back in 2016 so I would say it's too old.
It also hasn't been configured to use the crumb security (CSRF not enabled)

wimjongman commented:
> I would suggest merging the code and deal with any issues that crop up with a new ticket

Hey George. Cool! Do you want to take a look at the review comments before merging?

ruspl-afed reviewed (approved):
I've removed my comments regarding issues in .target, I need to have closer look later.

gnl42 commented:
> > I would suggest merging the code and deal with any issues that crop up with a new ticket
> 
> Hey George. Cool! Do you want to take a look at the review comments before merging?

I think I've addressed the comments to the best of my ability. When I tried to address the comments about the .target I ran into  unresolved references that I have no idea how to fix

gnl42 commented:
Don't merge yet. Started looking into the session Id problem and found some very not nice behaviour

gnl42 commented:
Solved the "sessionId" issue.

user/pw can run build now. Caveat is that if Jenkins is using StrictCrumbIssuer with a timeout set for the crumb one will be unable to run builds once the crumb expires. I can't see any way to have a new crumb created in this situation with how the build framework is designed.
Does not happen with a token, the recommended way, so I'm ok with this behaviour.

wimjongman commented:
George, is the token used in the password field?

gnl42 commented:
Hi Wim,
   yes.

It's getting complicated. Depending on the system, authentication can be by:

  * Token
  * username/pw
  * username/token

George


On 2023-03-07 00:30, Wim Jongman wrote:
>
> George, is the token used in the password field?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/eclipse-mylyn/org.eclipse.mylyn/pull/60#issuecomment-1457755349>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AP36HKKWKYQEAMK4YEUUM23W23WZDANCNFSM6AAAAAAVFIKDPQ>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

wimjongman commented:
Yes, it is. To be clear by 'system' do you mean systems like Jenkins or different Jenkins versions? Is the UI still using "Hudson"?

gnl42 commented:
Yes,

  * Token: Bamboo 8, GitHub, Jira 8
  * username/pw: Bamboo 8, Jenkins, Jira 8
  * username/token: Jira 10, Jenkins (modern)

The connector is advertised as 'Hudson (supports Jenkins)' but I don't 
see anything in the UI itself


On 2023-03-07 01:21, Wim Jongman wrote:
>
> Yes, it is. To be clear by 'system' do you mean systems like Jenkins 
> or different Jenkins versions? Is the UI still using "Hudson"?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/eclipse-mylyn/org.eclipse.mylyn/pull/60#issuecomment-1457826938>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AP36HKKAE3NTNRANAN6QT6DW234ZTANCNFSM6AAAAAAVFIKDPQ>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

BeckerFrank commented:
> @BeckerFrank
> 
> > Maybe the versions are to old to support that or we have never configure this.
> 
> Looks like 2.32.3 was released back in 2016 so I would say it's too old. It also hasn't been configured to use the crumb security (CSRF not enabled)
Yes we should use a newer Jenkins version in our new test infrastructure (see Multipass setup supports >= 2_303_3)

gnl42 commented:
Could someone merge the PR?
